### PR TITLE
pythonPackages.nltk: 3.2.5 -> 3.4

### PR DESCRIPTION
**Pull request by couchemar**
###### Motivation for this change
Update to the most recent version. 

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Review by matthiasbeyer — APPROVED** [review]

**Comment by lheckemann** [comment]
@GrahamcOfBorg build python3Packages.nltk

File: pkgs/development/python-modules/nltk/default.nix [hunk 1]
```diff
@@ -1,15 +1,16 @@
-{ fetchurl, buildPythonPackage, lib, six, pythonAtLeast, pythonOlder }:
+{ fetchurl, buildPythonPackage, lib, six, pythonAtLeast, pythonOlder
+, singledispatch }:
 
 buildPythonPackage rec {
-  version = "3.2.5";
+  version = "3.4";
   pname = "nltk";
 
   src = fetchurl {
```
Comment (dotlambda) [inline pkgs/development/python-modules/nltk/default.nix]:
```suggestion
  src = fetchPypi {
```

File: pkgs/development/python-modules/nltk/default.nix [hunk 2]
```diff
@@ -1,15 +1,16 @@
...
-    url = "mirror://pypi/n/nltk/nltk-${version}.tar.gz";
-    sha256 = "2661f9971d983db314bbebd51ba770811a362c6597fd0f303bb1d3beadcb4834";
+    url = "mirror://pypi/n/nltk/nltk-${version}.zip";
+    sha256 = "286f6797204ffdb52525a1d21ec0a221ec68b8e3fa4f2d25f412ac8e63c70e8d";
   };
 
-  propagatedBuildInputs = [ six ];
+  propagatedBuildInputs = [ singledispatch six ];
 
   disabled = pythonOlder "2.7" || pythonOlder "3.4" && (pythonAtLeast "3.0");
```
Comment (dotlambda) [inline pkgs/development/python-modules/nltk/default.nix]: This is no longer needed.

File: pkgs/development/python-modules/nltk/default.nix [hunk 3]
```diff
@@ -1,15 +1,16 @@
-{ fetchurl, buildPythonPackage, lib, six, pythonAtLeast, pythonOlder }:
+{ fetchurl, buildPythonPackage, lib, six, pythonAtLeast, pythonOlder
```
Comment (dotlambda) [inline pkgs/development/python-modules/nltk/default.nix]:
```suggestion
{ fetchPypi, buildPythonPackage, lib, six
```

**Review by dotlambda — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/nltk/default.nix (hunk 2 above)
Comment (couchemar) [reply pkgs/development/python-modules/nltk/default.nix]: Partly true. we need to patch `setup.py` to remove this dependency for python3.4+ 
https://github.com/nltk/nltk/blob/3.4/setup.py#L98

**Comment by couchemar** [comment]
@dotlambda https://github.com/NixOS/nixpkgs/pull/59204/commits/ed004dced5395f9f06b283882576e8446ae74e84 check fixes

File: pkgs/development/python-modules/nltk/default.nix (hunk 2 above)
Comment (dotlambda) [reply pkgs/development/python-modules/nltk/default.nix]: I was referring to `disabled = ...` line, which you removed :+1:

File: pkgs/development/python-modules/nltk/default.nix [hunk 4]
```diff
@@ -1,17 +1,21 @@
...
   };
 
-  propagatedBuildInputs = [ six ];
+  propagatedBuildInputs = [ six ] ++ lib.optional (!isPy3k) singledispatch;
 
-  disabled = pythonOlder "2.7" || pythonOlder "3.4" && (pythonAtLeast "3.0");
+  # TODO: remove patch during update to new version
+  patchPhase = ''
+  sed -ie 's/singledispatch/singledispatch; python_version < \\"3.4\\"/g' setup.py
+  '';
```
Comment (dotlambda) [inline pkgs/development/python-modules/nltk/default.nix]: Please use `fetchpatch` with https://github.com/nltk/nltk/commit/3966111cbf2f35fb86082b2f12acd90d75e9b8bb.patch.

**Review by dotlambda — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/nltk/default.nix (hunk 4 above)
Comment (couchemar) [reply pkgs/development/python-modules/nltk/default.nix]: Do we really need this? What a point?

Comment (matthiasbeyer) [reply pkgs/development/python-modules/nltk/default.nix]: It is way more robust than a sed. If the patch does not apply, the build fails. If your sed breaks something, we end up with a strange error output to debug - or worse a half-working package.

Comment (couchemar) [reply pkgs/development/python-modules/nltk/default.nix]: Got it. I'll fix soon. Thank you.

Comment (couchemar) [reply pkgs/development/python-modules/nltk/default.nix]: @dotlambda @matthiasbeyer https://github.com/NixOS/nixpkgs/pull/59204/commits/2e5aceadb86c32f391708930e6ec17d5076825ea Please check

**Comment by matthiasbeyer** [comment]
Nice. Now you can rebase-squash and expand the commit message to tell what you did! 

Awesome work!

**Review by matthiasbeyer — APPROVED** [review]
Patch looks good to me.

**Comment by lheckemann** [comment]
@GrahamcOfBorg build python3Packages.nltk

File: pkgs/development/python-modules/nltk/default.nix [hunk 5]
```diff
@@ -1,17 +1,25 @@
...
   };
 
-  propagatedBuildInputs = [ six ];
+  propagatedBuildInputs = [ six ] ++ lib.optional (!isPy3k) singledispatch;
 
-  disabled = pythonOlder "2.7" || pythonOlder "3.4" && (pythonAtLeast "3.0");
+  # TODO: remove patch during update to new version
+  patches = [
+    (fetchpatch {
+      url = "https://github.com/nltk/nltk/commit/3966111cbf2f35fb86082b2f12acd90d75e9b8bb.patch";
```
Comment (dotlambda) [inline pkgs/development/python-modules/nltk/default.nix]:
```suggestion
      url = https://github.com/nltk/nltk/commit/3966111cbf2f35fb86082b2f12acd90d75e9b8bb.patch;
```

**Review by dotlambda — CHANGES_REQUESTED** [review]
LGTM, minor nitpick

**Comment by dotlambda** [comment]
@GrahamcOfBorg build python2.pkgs.nltk python3.pkgs.nltk

**Pull request merged**